Adds Center Guides

radiovisual opened the pull request:
This PR adds center guides to the rulers. You can toggle the visibility of the center guides on or off with the keyboard shortcut <kbd>⌘;</kbd> (which is the same shortcut for toggling guide visibility in Photoshop), or you can use the `Toggle Center Guides` option in the `lr` menu. 

The center guides are not visible by default.

There are also a few small formatting tweaks. To get the center guides to find the exact center of the ruler space, the margins on the top, right, bottom and left needed to be consistent, so a few of the UI elements on the ruler needed to change location slightly. This is what the tweaked ruler looks like (when the center guides are active):

![center-guides](https://cloud.githubusercontent.com/assets/5614571/14067978/0742460e-f46d-11e5-957f-71e724a770d6.png)

radiovisual committed:
adds center guides

radiovisual committed:
add center-guide shortcut to help page

radiovisual committed:
only the focused ruler receives the center-guide shortcut.

radiovisual commented:
My latest commit: [only the focused ruler receives the center-guide shortcut](https://github.com/mikaa123/linear/pull/17/commits/29ddbbdd8f7916a9fe29bc16be102eb96c676938) ensures that the shortcut <kbd>⌘;</kbd> only works on the **active** (focused) ruler, and not **all** the ruler windows (as it did previously). This is much better because it lets the users decided which rulers should show the center guides, instead of forcing them to choose between "_on for every ruler_", or "_off for every ruler_".

radiovisual committed:
simple linting pass for code consistency

radiovisual commented on main.js:
```diff
@@ -15,7 +15,6 @@ const dataStore = require('./src/data-store');
 // });
 
 let rulers = [];
-let mainWindow;
```
This was removed, because it was not used anywhere in the file/project.

radiovisual commented on main.js:
```diff
@@ -28,13 +27,13 @@ let helpWindow;
+		'transparent': true,
+		'alwaysOnTop': true,
+		'x': _windowInfo.x,
+		'y': _windowInfo.y,
```
linters like it when property names are consistent in their quoting (either all quotes or no quotes, not a combination of both)

radiovisual commented on main.js:
```diff
@@ -114,62 +113,62 @@ function toggleRulerCommand() {
+			label: 'Settings',
+			click: () => {
+				if (settingsWindow) {
+					return;
```
Just ensured that the spacing/tabs were consistent in the whole file.

radiovisual commented on main.js:
```diff
@@ -196,7 +195,7 @@ app.on('ready', function() {
 
 	// We observe changes on the `rulers` array to enable/disable the toggle menu.
-	Array.observe(rulers, () => toggleRulersMenu.enabled = !!rulers.length);
+	Array.observe(rulers, () => toggleRulersMenu.enabled = Boolean(rulers.length));
```
This new syntax won't change the behavior, but [removes unnecessary complexity or erroneous type checking](http://eslint.org/docs/rules/no-extra-boolean-cast), and makes the line more readable.

radiovisual commented:
Sorry about mixing the "linting" pass in with this PR. I should have broken them up into two separate Pull Requests. I just starting making the spaces / tabs more consistent and got carried away. :sweat_smile:

mikaa123 commented:
Wow that was fast! Really cool :)
I pulled your branch and ran a few tests. It works great! I show the center guides on the ruler I want without impacting others, which is really nice.

I don't have any problem with reorganizing the ruler's layout, however there's a little problem with small rulers: due to the min-width of the window, we can't make them smaller (and measure small values). It's not possible to go under 50px width and 25 height.
![little ruler screenshot](https://cloud.githubusercontent.com/assets/428280/14086539/dc650a98-f525-11e5-849b-3551c8d35ca6.png)

Got any idea on what we could do to fix that?
Fantastic work btw!

mikaa123 commented on src/app/ruler/ruler.css:
```diff
@@ -122,3 +121,28 @@ body {
+	left: 50%;
+	height: 100%;
+	width: 1px;
+	border-right: dashed red 1px;
```
Not personally a big fan of the red haha, maybe in white/gray or lighter red? Don't worry about it though, we'll find one haha

mikaa123 commented on src/app/ruler/ruler.js:
```diff
@@ -146,5 +153,18 @@ document.querySelector('.ruler__inner__theme').addEventListener('click', () => {
+function toggleCenterGuides() {
+	showCenterGuides = !showCenterGuides;
+
+	if (showCenterGuides === true) {
```
Do we have to check for `true`, or would `truthy` be enough here?
e.g. `if (showCenterGuides)`

radiovisual replied:
"truthy" is enough, I just wanted to make sure that the intention of the code was super readable while I was working the logic out in my head, I will update statements like this to be more concise (truthy).

radiovisual commented:
Yes, I think I know the reason why the small rulers aren't possible, I will take a look at the code later on and fix it. As for the color red, I can change the color to be a light gray, but what do you think about letting the users set the color for the guides, and their color preferences get saved in the `.linear` file?

mikaa123 commented:
@radiovisual I think for now we should just hard-code the colors (I'll ask my designer friends for advices!). Later, if we allow custom CSS (themes) then they'd be able to change the color. So don't worry too much about finding the right color ;)

radiovisual commented:
Ok, no problem. I will leave it red until you let me know what color values you prefer.

radiovisual committed:
allow small rulers

radiovisual commented:
Ok, so it was easy to fix the "small rulers" problem. All that needed to be done was set the window `min-width` and `min-height` to `101px`, which is the value that would allow the ruler to be a size of `(0, 0)` (pictured below):

![small-rulers](https://cloud.githubusercontent.com/assets/5614571/14088434/75b517f0-f527-11e5-885a-e8d175ea5161.png)

radiovisual commented:
but now there is some uneven padding on the right and left, so I will fix that now....

radiovisual committed:
truthy shall suffice

radiovisual committed:
lightgray center guides

radiovisual commented:
Ok, so I have changed the guides to default to lightgray for now, and I also fixed the problem where small rulers were getting weird padding behaviors on the right side. This is the result:

**Lightgray Center Guides:**
![lightgray-guides-new-borders](https://cloud.githubusercontent.com/assets/5614571/14088952/3386aeb8-f52a-11e5-8134-5beb5b8fb0d1.png)

**Borderless rulers allow for proper padding:**
![new-small-size](https://cloud.githubusercontent.com/assets/5614571/14088951/3384aad2-f52a-11e5-8933-6190192943c3.png)

radiovisual committed:
center guides dark theme

radiovisual commented:
Ok, now the center guides are supported in the dark theme as well. When you merge this PR you can always change the default colors, but here is what they are as of right now:

**Dark Theme & Light Theme Center Guides**:
![dark-light-themes](https://cloud.githubusercontent.com/assets/5614571/14089747/67d850aa-f52e-11e5-9472-f4a52ec9a09e.png)

mikaa123 commented on main.js:
```diff
@@ -52,6 +51,14 @@ function createNewRuler(windowInfo) {
+ */
+function toggleCenterGuidesCommand() {
+	var activeRuler = BrowserWindow.getFocusedWindow();
+	activeRuler.send('toggle-center-guides');
```
Let's make sure `activeRuler` isn't undefined before calling `send`. Otherwise an error is thrown when no ruler is selected (for example when creating a ruler, clicking on another application, then going on the menu and clicking on Toggle Center Guides)

radiovisual replied:
Fixed!

mikaa123 commented on main.js:
```diff
@@ -106,56 +113,62 @@ function toggleRulerCommand() {
+			label: 'Toggle Center Guides',
+			accelerator: 'Command+;',
+			click: toggleCenterGuidesCommand,
+			enabled: true
```
Do you know a way we could hide this menu if no Ruler has focus? That would be really awesome!

radiovisual replied:
It would be possible, if we scaffolded out a state-machine to pub-sub to specific app-level events and then redraw the menu accordingly. IMO, a state-machine would only be justified if we were to build out more features that could benefit from subscribing to state changes (like various react components, or more interactive features embedded into the ruler), so we could put a feature like that on the V2 roadmap if a state machine makes sense.

radiovisual replied:
I just looked into it, and perhaps there is an easier way. I will experiment.

radiovisual replied:
false alarm. my original thinking is correct, it would require us to make a state machine and rebuild the menu based on the pub-sub of various app-level events. Its a cool idea, but the re-write would take us beyond the scope of this PR.

mikaa123 replied:
Yeah that's what I was also thinking :(
Let's start thinking about that pub/sub architecture for later!

I'll merge this later today. :)

radiovisual committed:
protect activeBrowser against undefined

mikaa123 commented:
I ran into the following issue:
- Start the app
- A new ruler pops
- Click on it (so it has the focus)
- Lr icon > Toggle Center Guide won't toggle (`BrowserWindow.getFocusedWindow()` returns undefined)

It's kind of weird because sometimes it works, when there are a few rulers! I'll look into that a little bit later

radiovisual commented:
I think its an issue with Electron. Yesterday I was experimenting with the `BrowserWindow.getFocusedWindow()` in relation to the `blur` and `focus` events and even inside the `focus` handler `getFocusedWindow()` would return `undefined`, but that was only in my experiments for disabling the context menu "Toggle Center Guides", I haven't had any problems with this Center Guides PR, so I will do this:
1. I will review my code to make sure I didn't introduce any bugs, and
2. If I find more strange problems with the `getFocusedWindow()` I will open an issue with Electron

radiovisual committed:
cleanup

radiovisual commented:
When I run my local tests, it all works for me without any issue. So its weird how the `getFocusedWindow()` function is not playing very nicely. 

I am going to setup an issue with electron based off of my experiments yesterday, and see if that will help  us learn anything new.

mikaa123 commented:
Thanks! You did an amazing job. Here is what I'll do: I'll merge your PR and change `BrowserWindow.getFocusedWindow()` with:
- Have a `lastFocusedRuler` variable
- Anytime a ruler is created or focused, have `lastFocusedRuler` point to it

That way, no more problem with `BrowserWindow.getFocusedWindow()` :) Is that fine with you?

radiovisual commented:
Yeah, if that works for you, then go for it! If there are any new problems, we will fix them! :smile:

mikaa123 commented:
Mergeeeed!

radiovisual commented:
Just for future reference: I [opened an issue with Electron](https://github.com/atom/electron/issues/4942), to see what is going on with `getFocusedWindow()`

zulhfreelancer commented:
NICE!

radiovisual commented:
@mikaa123 , just for future reference, [here is the current workaround](https://github.com/atom/electron/issues/984#issuecomment-69233279) for the `.getFocusedWindow()` problem we were having (it is a known issue with Electron).

The current workaround is to simply wrap the functionality in `setImmediate()`. I just ran my own test, and this setup works perfectly:

``` js
win.on('blur', blurHandler);
win.on('focus', blurHandler);
// ...
function blurHandler() {
    setImmediate(function() {
        let focusedWindow = BrowserWindow.getFocusedWindow();

        if (focusedWindow) {
            console.log('focus!');
        } else {
            console.log('blur!');
        }
    });
}
```
